src/backend/distributed/metadata/metadata_sync.c

```diff
@@ -3954,47 +3814,491 @@ ColocationGroupDeleteCommand(uint32 colocationId)
+		return;
+	}
+
+	/* send commands to new workers, the current user should be a superuser */
+	Assert(superuser());
+
+	if (context->transactionMode == METADATA_SYNC_TRANSACTIONAL)
+	{
+		Use2PCForCoordinatedTransaction();
+		List *workerNodes = context->activatedWorkerNodeList;
+		SendMetadataCommandListToWorkerListInCoordinatedTransaction(workerNodes,
+																	CurrentUserName(),
+																	commands);
+	}
+	else if (context->transactionMode == METADATA_SYNC_NON_TRANSACTIONAL)
+	{
+		List *workerConnections = context->activatedWorkerBareConnections;
+		SendCommandListToWorkerListWithBareConnections(workerConnections, commands);
+	}
+	else
+	{
+		pg_unreachable();
+	}
+}
+
+
+/*
+ * SendOrCollectCommandListToMetadataNodes sends the commands to the metadata nodes with
+ * bare connections inside metadatacontext or via coordinated connections.
+ * Note that when context only collects commands, we add commands into the context
+ * without sending the commands.
+ */
+void
+SendOrCollectCommandListToMetadataNodes(MetadataSyncContext *context, List *commands)
+{
+	/*
+	 * do not send any command to workers if we collcet commands.
+	 * Collect commands into metadataSyncContext's collected command
+	 * list.
+	 */
+	if (MetadataSyncCollectsCommands(context))
+	{
+		context->collectedCommands = list_concat(context->collectedCommands, commands);
+		return;
+	}
+
+	/* send commands to new workers, the current user should be a superuser */
+	Assert(superuser());
+
+	if (context->transactionMode == METADATA_SYNC_TRANSACTIONAL)
+	{
+		Use2PCForCoordinatedTransaction();
+		List *metadataNodes = TargetWorkerSetNodeList(NON_COORDINATOR_METADATA_NODES,
+													  RowShareLock);
+		SendMetadataCommandListToWorkerListInCoordinatedTransaction(metadataNodes,
+																	CurrentUserName(),
+																	commands);
+	}
+	else if (context->transactionMode == METADATA_SYNC_NON_TRANSACTIONAL)
+	{
+		SendBareCommandListToMetadataWorkers(commands);
+	}
+	else
+	{
+		pg_unreachable();
+	}
+}
+
+
+/*
+ * SendOrCollectCommandListToSingleNode sends the commands to the specific worker
+ * indexed by nodeIdx with bare connection inside metadatacontext or via coordinated
+ * connection. Note that when context only collects commands, we add commands into
+ * the context without sending the commands.
+ */
+void
+SendOrCollectCommandListToSingleNode(MetadataSyncContext *context, List *commands,
+									 int nodeIdx)
+{
+	/*
+	 * Do not send any command to workers if we collect commands.
+	 * Collect commands into metadataSyncContext's collected command
+	 * list.
+	 */
+	if (MetadataSyncCollectsCommands(context))
+	{
+		context->collectedCommands = list_concat(context->collectedCommands, commands);
+		return;
+	}
+
+	/* send commands to new workers, the current user should be a superuser */
+	Assert(superuser());
+
+	if (context->transactionMode == METADATA_SYNC_TRANSACTIONAL)
+	{
+		Use2PCForCoordinatedTransaction();
```

Comment:
maybe we should move Use2PCForCoordinatedTransaction() to `CreateMetadataSyncContext`, otherwise it is called too many times